ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -376,19 +379,66 @@ static int HandleFileOpOperation(
     atomic_fetch_add(&s_numActiveKauthEvents, 1);
     
     vfs_context_t context = vfs_context_create(NULL);
+    vnode_t currentVnodeFromPath = NULLVP;
 
-    if (KAUTH_FILEOP_CLOSE == action)
+    if (KAUTH_FILEOP_RENAME == action ||
+        KAUTH_FILEOP_LINK == action)
     {
-        vnode_t currentVnode = reinterpret_cast<vnode_t>(arg0);
-        // arg1 is the (const char *) path
-        int closeFlags = static_cast<int>(arg2);
+        // arg0 is the (const char *) fromPath (or the file being linked to)
+        const char* newPath = (const char*)arg1;
         
-        vtype vnodeType = vnode_vtype(currentVnode);
-        if (ShouldIgnoreVnodeType(vnodeType, currentVnode))
+        // TODO(Mac): Improve error handling
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnodeFromPath, context);
+        if (0 != toErr)
         {
             goto CleanupAndReturn;
         }
         
+        VirtualizationRoot* root = nullptr;
+        int pid;
+        if (!ShouldHandleFileOpEvent(
+                context,
+                currentVnodeFromPath,
+                action,
+                &root,
+                &pid))
+        {
+            goto CleanupAndReturn;
+        }
+        
+        char procname[MAXCOMLEN + 1];
+        proc_name(pid, procname, MAXCOMLEN + 1);
+
+        MessageType messageType;
+        if (KAUTH_FILEOP_RENAME == action)
+        {
+            messageType = vnode_isdir(currentVnodeFromPath) ? MessageType_KtoU_NotifyDirectoryRenamed : MessageType_KtoU_NotifyFileRenamed;
+        }
+        else
+        {
+            messageType = MessageType_KtoU_NotifyFileHardLinkCreated;
+        }
+
+        int kauthResult;
+        int kauthError;
+        if (!TrySendRequestAndWaitForResponse(
+            root,
```

Comment:
For consistency with your other code, indent one more level here

Reply:
:+1: